sdk/core/core/inc/az_json.h

```diff
@@ -182,81 +196,233 @@ AZ_NODISCARD az_result az_json_token_get_string(az_json_token const* token, az_s
 
-/*
- * @brief az_json_builder_init initializes an az_json_builder which writes JSON into a buffer.
+/**
+ * @brief Initializes an #az_json_builder which writes JSON text into a buffer.
  *
- * @param json_builder A pointer to an az_json_builder instance to initialize.
- * @param json_buffer An az_span over the buffer where th JSON object is to be written.
- * @return AZ_OK if the az_json_builder is initialized correctly.
+ * @param[out] json_builder A pointer to an #az_json_builder instance to initialize.
+ * @param[in] destination_buffer An #az_span over the byte buffer where the JSON text is to be
+ * written.
+ * @param[in] options __[nullable]__ A reference to an #az_json_builder_options
+ * structure which defines custom behavior of the #az_json_builder. If `NULL` is passed, the builder
+ * will use the default options (i.e. #az_json_builder_options_default()).
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if the az_json_builder is initialized successfully
  */
-AZ_NODISCARD AZ_INLINE az_result
-az_json_builder_init(az_json_builder* json_builder, az_span json_buffer)
+AZ_NODISCARD AZ_INLINE az_result az_json_builder_init(
+    az_json_builder* json_builder,
+    az_span destination_buffer,
+    az_json_builder_options const* options)
 {
   *json_builder
-      = (az_json_builder){ ._internal = { .json = json_buffer, .length = 0, .need_comma = false } };
+      = (az_json_builder){ ._internal = {
+                               .destination_buffer = destination_buffer,
+                               .bytes_written = 0,
+                               .need_comma = false,
+                               .token_kind = AZ_JSON_TOKEN_NONE,
+                               .bit_stack = { 0 },
+                               .options
+                               = options == NULL ? az_json_builder_options_default() : *options,
+                           } };
   return AZ_OK;
 }
 
-/*
- * @brief az_json_builder_span_get returns the az_span containing the final JSON object.
+/**
+ * @brief Returns the #az_span containing the JSON text written to the underlying buffer so far.
+ *
+ * @param[in] json_builder A pointer to an #az_json_builder instance wrapping the destination
+ * buffer.
+ *
+ * @note Do NOT modify or override the contents of the returned #az_span unless you are no longer
+ * building JSON text into it.
  *
- * @param json_builder A pointer to an az_json_builder instance wrapping the JSON buffer.
- * @return an az_span containing the final JSON object.
+ * @return An #az_span containing the JSON text built so far.
  */
-AZ_NODISCARD AZ_INLINE az_span az_json_builder_span_get(az_json_builder const* json_builder)
+AZ_NODISCARD AZ_INLINE az_span az_json_builder_get_json(az_json_builder const* json_builder)
 {
-  return az_span_slice(json_builder->_internal.json, 0, json_builder->_internal.length);
+  // TODO: Consider precondition validation of the state before returning.
```

Comment:
I wouldn't pre-condition returning only a valid json.
As a customer I might want to create something that I want to use latter to have it add it multiple times into another json, like
```c
   // use builder to create one buffer with az_span re-use = ` "someOption": true, "otherOption": `
  // one json
  az_json_append(json, re-use);
  az_json_append(json, false);
  // json 2
  az_json_append(json, re-use);
  az_json_append(json, true);
  //  Now I have twot jsons:
         // {"someOption": true, "otherOption": false} and {"someOption": true, "otherOption": true}
```

Making invalid to return WIP json from the builder would require customer to get the buffer from internal to re-use it as a work-around.

Reply:
> If this is easy to do in the debugger, then the precondition is probably OK

If at dev/debug time, the caller doesn't modify the infinite sleep precondition OR change it to an undefined behavior like null dereference, then like any usage of precondition, it will be hard to debug and see partial JSON in the debugger.

However, since it is dev/debug time, potentially the user can override the precondition to no-op.

Definitely something to consider. Will go with the safer, strict route for now and adjust later.

Reply:
Victor, I am leaning towards keeping the stricter condition and only revert that once we have a concrete use case and scenario.

Regarding the example you gave:
1) Why wouldn't the caller keep passing the same instance of the builder (or create multiple) rather than trying to reuse the same underlying buffer. That is error prone and seems suspect.
2) If the caller get's the json (i.e. slice of the underlying buffer) and then continues to use the same builder for writing other stuff, they will override the data and cause data corruption. That is why there is a warning in the `@remarks` of the docs. Sharing/reusing span slices like this could be problematic.

We could add a "reset" method that clears the state of the json builder to allow your use case, and even in your example, the two JSON snippets are complete and well-formed, so the caller can already do that.